recstudio/model/mf/dice.py

```diff
@@ -0,0 +1,213 @@
+import torch
+import numpy as np
+from recstudio.ann.sampler import Sampler
+from recstudio.model.mf.bpr import BPR
+from recstudio.model import basemodel, loss_func
+import time
+
+class DICE(BPR):               
+        
+    def add_model_specific_args(parent_parser):
+        parent_parser = basemodel.Recommender.add_model_specific_args(parent_parser)
+        parent_parser.add_argument_group('DICE')
+        parent_parser.add_argument("--dis_loss", type=str, default='l1', help='discrepency loss function')
+        parent_parser.add_argument("--int_weight", type=float, default=0.1, help='weight for interest term in the loss function')
+        parent_parser.add_argument("--pop_weight", type=float, default=0.1, help='weight for popularity term in the loss function')
+        parent_parser.add_argument("--dis_pen", type=float, default=0.01, help='discrepency penalty')
+        parent_parser.add_argument("--margin_up", type=float, default=40.0, help='margin for negative but more popular sampling')
+        parent_parser.add_argument("--margin_down", type=float, default=40.0, help='margin for negative and less popular sampling')
+        parent_parser.add_argument("--pool", type=int, default=40, help='pool for negative sampling')
+        parent_parser.add_argument("--adaptive", type=bool, default=True, help='adapt hyper-parameters or not')
+        parent_parser.add_argument("--margin_decay", type=float, default=0.9, help='decay of margin')
+        parent_parser.add_argument("--loss_decay", type=float, default=0.9, help='decay of loss')
+        return parent_parser             
+
+    def _get_query_encoder(self, train_data):
+        int = torch.nn.Embedding(train_data.num_users, self.embed_dim, padding_idx=0)
+        pop = torch.nn.Embedding(train_data.num_users, self.embed_dim, padding_idx=0)
+        class DICEQueryEncoder(torch.nn.Module):
```

Comment:
The int and pop in query encoder could be defined as one Embedding, with dimension 2*self.embed_dim?

```suggestion
        torch.nn.Embedding(train_data.num_users, 2*self.embed_dim, padding_idx=0)
```